### DISCUSSION TITLE
DAI 2023-2024 - Java IOs benchmarking - Valentin Ricard

### DISCUSSION THREAD

**Discussion by knownasred** [heig-vd-dai-course/heig-vd-dai-course#67] [Show and tell]
[Link to the repository](https://github.com/valentin-ricard/java-io-lab/) - I promise, the readme has interesting details for that totally overengineered solution ^^
## Results
| Benchmark                | File Size | Operations per Second (Error margin) |
|--------------------------|-----------|---------------------------------------|
| BinaryFileReader         | 1         | 13742,010 ± 19242,016                 |
|                          | 1Ko       | 274,237 ± 432,215                     |
|                          | 1Mo       | 0,258 ± 0,204                         |
|                          | 5Mo       | 0,056 ± 0,044                         |
| BinaryFileWriter         | 1         | 1419,634 ± 1445,961                   |
|                          | 1Ko       | 88,015 ± 85,638                       |
|                          | 1Mo       | 0,234 ± 0,292                         |
|                          | 5Mo       | 0,051 ± 0,015                         |
| BufferedBinaryFileReader | 1         | 15082,600 ± 24151,246                 |
|                          | 1Ko       | 10703,042 ± 6882,253                  |
|                          | 1Mo       | 40,761 ± 120,443                      |
|                          | 5Mo       | 8,508 ± 11,231                        |
| BufferedBinaryFileWriter | 1         | 1235,167 ± 1690,709                   |
|                          | 1Ko       | 1150,627 ± 1627,975                   |
|                          | 1Mo       | 33,505 ± 30,731                       |
|                          | 5Mo       | 8,184 ± 4,309                         |
| BufferedTextFileReader   | 1         | 13385,310 ± 19776,565                 |
|                          | 1Ko       | 9748,486 ± 10160,881                  |
|                          | 1Mo       | 72,392 ± 35,745                       |
|                          | 5Mo       | 12,637 ± 6,261                        |
| BufferedTextFileWriter   | 1         | 1227,675 ± 2981,563                   |
|                          | 1Ko       | 1049,948 ± 1175,990                   |
|                          | 1Mo       | 33,110 ± 33,677                       |
|                          | 5Mo       | 6,303 ± 4,975                         |
| TextFileReader           | 1         | 11955,754 ± 4923,005                  |
|                          | 1Ko       | 8154,226 ± 5924,808                   |
|                          | 1Mo       | 38,364 ± 49,598                       |
|                          | 5Mo       | 7,783 ± 7,304                         |
| TextFileWriter           | 1         | 1246,487 ± 734,538                    |
|                          | 1Ko       | 1035,414 ± 363,519                    |
|                          | 1Mo       | 17,658 ± 16,241                       |
|                          | 5Mo       | 3,670 ± 0,627                         |

### Which type of stream is the most efficient for each use case?
The buffered versions are more performant.
### Why is it more efficient than the other types of streams?
They reduce the system calls necessary to get the contents (instead of a syscall per byte, a syscall for each block)
### What is the difference between binary data and text data?
Text data is variable width (you can have a character that requires multiple bytes to get access to.
### What is a character encoding?
A character encoding is a methodology to transpose characters to a binary encoding.
### Why is this methodology important?
This allows to check for performance using real-world data, instead of a theoretical one. I have a great example of this in the use of SIMD (Single Instruction Multiple Data) instructions on modern CPUs. In some cases, you can lose performance using the preparation of data for SIMD, instead of applying the instruction multiple times on different registers.

**Comment by ludelafo** [comment]
Well, I don't have much to say haha

I'm glad you were able to set up such a great benchmarking system (even though it took a long time to generate all the data) and that you took the time to go further regarding this topic!

Out of curiosity, was it the first time you set up a benchmarking system as displayed in your repo or have you done it before? Did it take long to understand and set up?

**Reply by knownasred** [reply]
For Java itself, this is a first time, but a lot of the parts of the frameworks are either the same as in other languages (in my case Rust via [Criterion](https://docs.rs/criterion/latest/criterion/)), or with other common patterns in java, for example the Dependency Injection pattern from SpringBoot, or the abuse of annotations from Spring Security / Spring Aspect-Oriented Programming.

The most difficult part is to work around the multithreaded aspect of the framework (so assigning different files), and making sure that the cleanup doesn't impact the performance.

**Reply by ludelafo** [reply]
Thank you for all the insights! I'm learning something new everyday, thank you!
